gitpod-io/gitpod
# [supervisor]: gitpod config changed analytics

## Description


This PR adds analytics for .gitpod.yml changes. We are track when changes are persisted to a disk and debounce them for 5s. After last change we analyse changes to top-level elements and report an event for each. It allows us to gain insights in what users are changing on top-level, e.g `ports` or `tasks`,   and avoid a lot of traffic at the same time. Later we can do more fine grained analysis to see which object properties, like `ports.name`.

I had to resolve couple issues which were discovered during review or were blocking testing:
- enable analytics env inside workspace in preview environments to verify with segment
- don't fire change events for syntactically broken .gitpod.yml files, thank you @andreafalzetti for noticing it
- adding git hook which keeps goland types up to date with gitpod schema, otherwise we will miss new properties, thank you @mustard-mh for brining forward that yaml parser only see properties which are defined in goland

## Related Issue(s)

Fixes https://github.com/gitpod-io/gitpod/issues/6894

## How to test


- Open staging untrusted.
- Start a workspace, open .gitpod.yml yet, wait for 5s.
- You should not yet see any `gitpod_config_changed` events.
- Now edit .gitpoy.yml, then you finish wait for 5s.
- Now you should see `gitpod_config_changed` events.
- Edit .gitpod.yml to add some syntax errors and semantic changes, then you finish wait for 5s.
- You should not yet see any `gitpod_config_changed` events.
- Edit .gitpod.yml to fix syntax errors, then you finish wait for 5s.
- Now you should see `gitpod_config_changed` events for semantic changes.

Check for regressions:
- Add new ports and check there are detected in the ports view.

## Release Notes

```release-note
NONE
```

## Documentation


## Werft options:
